corehq/sql_db/connections.py

```diff
@@ -36,11 +37,62 @@ def _get_db_alias_or_none(enigne_id):
         return None
 
 
+_thread_locals = local()
+
+
+READ_FROM_CITUS_STANDBYS = 'READ_FROM_CITUS_STANDBYS'
+
+
+def allow_read_from_citus_standbys():
+    return getattr(_thread_locals, READ_FROM_CITUS_STANDBYS, False)
+
+
+class read_from_citus_standbys(ContextDecorator):
+    def __enter__(self):
+        setattr(_thread_locals, READ_FROM_CITUS_STANDBYS, True)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        setattr(_thread_locals, READ_FROM_CITUS_STANDBYS, False)
+
+
+def is_citus_db(connection):
+    """
+    :param connection: either a sqlalchemy connection or a Django cursor
+    """
+    res = connection.execute("SELECT 1 FROM pg_extension WHERE extname = 'citus'")
+    if res is None:
+        res = list(connection)
+    return bool(list(res))
+
+
+def is_citus(engine, citus_list=[], non_citus_list=[]):
```

Comment:
the changes to this function don't seem necessary anymore since this isn't called often enough to make the caching worth while

Reply:
The function `supports_offset` that depends on this seems to be called from UCR rows saving, which might be often enough, no?

Reply:
you mean `supports_upsert`? I guess that makes sense.

Reply:
Er, yes `supports_upsert`.